core/src/main/java/org/infinispan/CacheStream.java

```diff
@@ -0,0 +1,251 @@
+package org.infinispan;
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+/**
+ * A {@link Stream} that has additional operations to monitor or control behavior when used from a {@link Cache}.  Note that
+ * you may only use these additional methods on the CacheStream before any intermediate operations are performed as
+ * a {@link Stream} is returned from those methods.
+ *
+ * <p>Whenever the iterator or spliterator methods are used the user <b>must</b> close the {@link Stream}
+ * that the method was invoked on after completion of its operation.  Failure to do so may cause a thread leakage if
+ * the iterator or spliterator are not fully consumed.</p>
+ *
+ * <p>When using stream that is backed by a distributed cache these operations will be performed using remote
+ * distribution controlled by the segments that each key maps to.  All intermediate operations are lazy, even the
+ * special cases described in later paragraphs and are not evaluated until a final terminal operation is invoked on
+ * the stream.  Essentially each set of intermediate operations is shipped to each remote node where they are applied
+ * to a local stream there and finally the terminal operation is completed.  If this stream is parallel the processing
+ * on remote nodes is also done using a parallel stream.</p>
+ *
+ * <p>Parallel distribution is enabled by default for all operations except for {@link CacheStream#iterator()} &
+ * {@link CacheStream#spliterator()}.  Please see {@link CacheStream#sequentialDistribution()} and
+ * {@link CacheStream#parallelDistribution()}.  With this disabled only a single node will process the operation
+ * at a time (includes locally).</p>
+ *
+ * <p>Rehash aware is enabled by default for all operations which will provide guaranteed consistency for all operations
+ * except for {@link CacheStream#forEach(Consumer)}.  Please the method above for details about its consistency
+ * guarantees.  If you wish to disable rehash aware operations you can disable them by calling
+ * {@link CacheStream#disableRehashAware()} which should provide better performance for some operations.  The
+ * performance is most affected for the key aware operations {@link CacheStream#iterator()},
+ * {@link CacheStream#spliterator()}, {@link CacheStream#forEach(Consumer)}</p>
+ *
+ * <p>Some terminal operators are special in that they act like an intermediate iterator operation.  That is that
```

Comment:
s/terminal operators/intermediate operators/?
The "intermediate iterator operation" definition sounds too low-level, maybe some operations use a spliterator instead of an iterator. Wouldn't it be enough to say they're "stateful operations" are described in the `Stream` javadoc?